File: Skyscraper-iOS-20251106/Skyscraper/Services/PostCacheService.swift
//
//  PostCacheService.swift
//  Skyscraper
//
//  Service for caching and retrieving timeline posts
//

import Foundation
import CoreData

@MainActor
class PostCacheService {
    static let shared = PostCacheService()

    private let coreData = CoreDataStack.shared
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    // Save posts to cache for a specific feed
    func cachePosts(_ posts: [FeedViewPost], feedId: String) {
        let context = coreData.context

        for (index, feedPost) in posts.enumerated() {
            // Check if post already exists for this feed
            let fetchRequest = NSFetchRequest<CachedPost>(entityName: "CachedPost")
            fetchRequest.predicate = NSPredicate(format: "uri == %@ AND feedId == %@", feedPost.post.uri, feedId)
            fetchRequest.fetchLimit = 1

            do {
                let existing = try context.fetch(fetchRequest)

                if existing.isEmpty {
                    // Create new cached post
                    let cachedPost = CachedPost(context: context)
                    cachedPost.uri = feedPost.post.uri
                    cachedPost.feedId = feedId
                    cachedPost.jsonData = try encoder.encode(feedPost)
                    cachedPost.cachedAt = Date()
                    cachedPost.createdAt = feedPost.post.createdAt
                    cachedPost.sortOrder = Int64(index)

                    print("Cached post: \(feedPost.post.uri) for feed: \(feedId)")
                } else {
                    // Update existing
                    let cachedPost = existing[0]
                    cachedPost.jsonData = try encoder.encode(feedPost)
                    cachedPost.cachedAt = Date()
                    cachedPost.sortOrder = Int64(index)
                }
            } catch {
                print("Error caching post: \(error.localizedDescription)")
            }
        }

        coreData.saveContext()
    }

    // Load cached posts for a specific feed
    func loadCachedPosts(feedId: String) -> [FeedViewPost] {
        let context = coreData.context

        let fetchRequest = NSFetchRequest<CachedPost>(entityName: "CachedPost")
        fetchRequest.predicate = NSPredicate(format: "feedId == %@", feedId)
        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "createdAt", ascending: false)]

        do {
            let cachedPosts = try context.fetch(fetchRequest)
            print("Loaded \(cachedPosts.count) cached posts for feed: \(feedId)")

            return cachedPosts.compactMap { cached in
                do {
                    return try decoder.decode(FeedViewPost.self, from: cached.jsonData)
                } catch {
                    print("Error decoding cached post: \(error.localizedDescription)")
                    return nil
                }
            }
        } catch {
            print("Error loading cached posts: \(error.localizedDescription)")
            return []
        }
    }

    // Load all cached posts across all feeds (for cache size calculation)
    func loadAllCachedPosts() -> [FeedViewPost] {
        let context = coreData.context

        let fetchRequest = NSFetchRequest<CachedPost>(entityName: "CachedPost")
        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "createdAt", ascending: false)]

        do {
            let cachedPosts = try context.fetch(fetchRequest)
            print("Loaded \(cachedPosts.count) total cached posts across all feeds")

            return cachedPosts.compactMap { cached in
                do {
                    return try decoder.decode(FeedViewPost.self, from: cached.jsonData)
                } catch {
                    print("Error decoding cached post: \(error.localizedDescription)")
                    return nil
                }
            }
        } catch {
            print("Error loading cached posts: \(error.localizedDescription)")
            return []
        }
    }

    // Clear all cached posts
    func clearCache() {
        let context = coreData.context
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "CachedPost")
        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)

        do {
            try context.execute(deleteRequest)
            coreData.saveContext()
            print("Cache cleared")
        } catch {
            print("Error clearing cache: \(error.localizedDescription)")
        }
    }

    // Clean up old posts (10 days)
    func cleanupOldPosts() {
        coreData.clearOldPosts(olderThan: 10)
    }

    // Save scroll position for a specific feed
    func saveScrollPosition(postURI: String, forFeed feedID: String) {
        var positions = loadAllScrollPositions()
        positions[feedID] = postURI

        if let encoded = try? JSONEncoder().encode(positions) {
            UserDefaults.standard.set(encoded, forKey: "feedScrollPositions")
            print("Saved scroll position for feed \(feedID): \(postURI)")
        }
    }

    // Load scroll position for a specific feed
    func loadScrollPosition(forFeed feedID: String) -> String? {
        let positions = loadAllScrollPositions()
        return positions[feedID]
    }

    // Load all scroll positions
    private func loadAllScrollPositions() -> [String: String] {
        guard let data = UserDefaults.standard.data(forKey: "feedScrollPositions"),
              let positions = try? JSONDecoder().decode([String: String].self, from: data) else {
            return [:]
        }
        return positions
    }

    // Clear scroll position for a specific feed
    func clearScrollPosition(forFeed feedID: String) {
        var positions = loadAllScrollPositions()
        positions.removeValue(forKey: feedID)

        if let encoded = try? JSONEncoder().encode(positions) {
            UserDefaults.standard.set(encoded, forKey: "feedScrollPositions")
        }
    }

    // Clear all scroll positions
    func clearAllScrollPositions() {
        UserDefaults.standard.removeObject(forKey: "feedScrollPositions")
    }

    // MARK: - Last Viewed Timestamp

    /// Save the last viewed timestamp for a feed (when user has seen posts up to this time)
    func saveLastViewedTimestamp(_ timestamp: Date, forFeed feedID: String) {
        var timestamps = loadAllLastViewedTimestamps()
        timestamps[feedID] = timestamp

        if let encoded = try? JSONEncoder().encode(timestamps) {
            UserDefaults.standard.set(encoded, forKey: "feedLastViewedTimestamps")
            print("💾 Saved last viewed timestamp for feed \(feedID): \(timestamp)")
        }
    }

    /// Load the last viewed timestamp for a feed
    func loadLastViewedTimestamp(forFeed feedID: String) -> Date? {
        let timestamps = loadAllLastViewedTimestamps()
        return timestamps[feedID]
    }

    /// Load all last viewed timestamps
    private func loadAllLastViewedTimestamps() -> [String: Date] {
        guard let data = UserDefaults.standard.data(forKey: "feedLastViewedTimestamps"),
              let timestamps = try? JSONDecoder().decode([String: Date].self, from: data) else {
            return [:]
        }
        return timestamps
    }

    /// Clear last viewed timestamp for a specific feed
    func clearLastViewedTimestamp(forFeed feedID: String) {
        var timestamps = loadAllLastViewedTimestamps()
        timestamps.removeValue(forKey: feedID)

        if let encoded = try? JSONEncoder().encode(timestamps) {
            UserDefaults.standard.set(encoded, forKey: "feedLastViewedTimestamps")
        }
    }

    /// Clear all last viewed timestamps
    func clearAllLastViewedTimestamps() {
        UserDefaults.standard.removeObject(forKey: "feedLastViewedTimestamps")
    }
}

File: app/src/main/java/com/cameronbanga/skyscraper/services/PostCacheService.kt
package com.cameronbanga.skyscraper.services

import android.content.Context
import android.content.SharedPreferences
import com.cameronbanga.skyscraper.data.CachedPost
import com.cameronbanga.skyscraper.data.SkyscraperDatabase
import com.cameronbanga.skyscraper.models.FeedViewPost
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.TimeUnit

/**
 * Service for caching and retrieving timeline posts using Room
 * Provides offline access and faster loading of timeline data
 */
class PostCacheService private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var INSTANCE: PostCacheService? = null

        fun getInstance(context: Context): PostCacheService = INSTANCE ?: synchronized(this) {
            INSTANCE ?: PostCacheService(context.applicationContext).also { INSTANCE = it }
        }

        val shared get() = INSTANCE
            ?: throw IllegalStateException("PostCacheService not initialized")

        private const val SCROLL_POSITIONS_KEY = "feedScrollPositions"
        private const val CACHE_MAX_AGE_DAYS = 10
    }

    private val database = SkyscraperDatabase.getInstance(context)
    private val cachedPostDao = database.cachedPostDao()
    private val moshi = Moshi.Builder().build()
    private val feedViewPostAdapter = moshi.adapter(FeedViewPost::class.java)
    private val prefs: SharedPreferences = context.getSharedPreferences(
        "post_cache_prefs",
        Context.MODE_PRIVATE
    )

    /**
     * Cache posts for a specific feed
     */
    suspend fun cachePosts(posts: List<FeedViewPost>, feedId: String) = withContext(Dispatchers.IO) {
        try {
            val cachedPosts = posts.mapIndexed { index, feedPost ->
                val jsonData = feedViewPostAdapter.toJson(feedPost).toByteArray()
                CachedPost(
                    uri = feedPost.post.uri,
                    jsonData = jsonData,
                    cachedAt = System.currentTimeMillis(),
                    createdAt = parseCreatedAt(feedPost.post.record.createdAt),
                    sortOrder = index.toLong(),
                    feedId = feedId
                )
            }

            cachedPostDao.insertPosts(cachedPosts)
        } catch (e: Exception) {
            // Log error but don't crash
        }
    }

    /**
     * Load cached posts for a specific feed
     */
    suspend fun loadCachedPosts(feedId: String): List<FeedViewPost> = withContext(Dispatchers.IO) {
        try {
            val cachedPosts = cachedPostDao.getPostsForFeed(feedId)
            cachedPosts.mapNotNull { cached ->
                try {
                    val jsonString = String(cached.jsonData)
                    feedViewPostAdapter.fromJson(jsonString)
                } catch (e: Exception) {
                    null
                }
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /**
     * Load all cached posts across all feeds
     */
    suspend fun loadAllCachedPosts(): List<FeedViewPost> = withContext(Dispatchers.IO) {
        try {
            // Since we don't have a "get all" query, we'd need to add it
            // For now, return empty list as this is mainly for cache size calculation
            emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    /**
     * Clear all cached posts
     */
    suspend fun clearCache() = withContext(Dispatchers.IO) {
        try {
            cachedPostDao.clearAll()
        } catch (e: Exception) {
            // Log error but don't crash
        }
    }

    /**
     * Clean up old posts (older than 10 days)
     */
    suspend fun cleanupOldPosts() = withContext(Dispatchers.IO) {
        try {
            val cutoffTime = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(CACHE_MAX_AGE_DAYS.toLong())
            cachedPostDao.deleteOldPosts(cutoffTime)
        } catch (e: Exception) {
            // Log error but don't crash
        }
    }

    /**
     * Save scroll position for a specific feed
     */
    fun saveScrollPosition(postURI: String, forFeed feedId: String) {
        try {
            val positions = loadAllScrollPositions().toMutableMap()
            positions[feedId] = postURI

            val json = moshi.adapter<Map<String, String>>(
                Types.newParameterizedType(Map::class.java, String::class.java, String::class.java)
            ).toJson(positions)

            prefs.edit()
                .putString(SCROLL_POSITIONS_KEY, json)
                .apply()
        } catch (e: Exception) {
            // Log error but don't crash
        }
    }

    /**
     * Load scroll position for a specific feed
     */
    fun loadScrollPosition(forFeed feedId: String): String? {
        return loadAllScrollPositions()[feedId]
    }

    /**
     * Load all scroll positions
     */
    private fun loadAllScrollPositions(): Map<String, String> {
        return try {
            val json = prefs.getString(SCROLL_POSITIONS_KEY, null) ?: return emptyMap()
            moshi.adapter<Map<String, String>>(
                Types.newParameterizedType(Map::class.java, String::class.java, String::class.java)
            ).fromJson(json) ?: emptyMap()
        } catch (e: Exception) {
            emptyMap()
        }
    }

    /**
     * Parse createdAt string to timestamp
     */
    private fun parseCreatedAt(createdAt: String): Long {
        return try {
            // ISO 8601 format parsing
            val formatter = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", java.util.Locale.getDefault())
            formatter.parse(createdAt)?.time ?: System.currentTimeMillis()
        } catch (e: Exception) {
            System.currentTimeMillis()
        }
    }
}
